src/gui/MixerView.cpp

```diff
@@ -385,48 +387,121 @@ void MixerView::deleteChannel(int index)
-	int selLine = m_currentMixerLine->channelIndex();
-
-	// in case the deleted channel is soloed or the remaining
-	// channels will be left in a muted state
-	Engine::mixer()->clearChannel(index);
-
-	// delete the real channel
-	Engine::mixer()->deleteChannel(index);
-
-	// delete the view
-	chLayout->removeWidget(m_mixerChannelViews[index]->m_mixerLine);
-	m_racksLayout->removeWidget( m_mixerChannelViews[index]->m_rackView );
-	delete m_mixerChannelViews[index]->m_fader;
-	delete m_mixerChannelViews[index]->m_muteBtn;
-	delete m_mixerChannelViews[index]->m_soloBtn;
-	// delete mixerLine later to prevent a crash when deleting from context menu
-	m_mixerChannelViews[index]->m_mixerLine->hide();
-	m_mixerChannelViews[index]->m_mixerLine->deleteLater();
-	delete m_mixerChannelViews[index]->m_rackView;
-	delete m_mixerChannelViews[index];
-	m_channelAreaWidget->adjustSize();
-
-	// make sure every channel knows what index it is
-	for(int i=index + 1; i<m_mixerChannelViews.size(); ++i)
+	// if channel is "used", confirm that the user want to delete the channel anyway
+	if (confirmRemoval(index))
 	{
-		m_mixerChannelViews[i]->m_mixerLine->setChannelIndex(i-1);
+		// remember selected line
+		int selLine = m_currentMixerLine->channelIndex();
+
+		// in case the deleted channel is soloed or the remaining
+		// channels will be left in a muted state
+		Engine::mixer()->clearChannel(index);
+
+		// delete the real channel
+		Engine::mixer()->deleteChannel(index);
+
+		// delete the view
+		chLayout->removeWidget(m_mixerChannelViews[index]->m_mixerLine);
+		m_racksLayout->removeWidget(m_mixerChannelViews[index]->m_rackView);
+		delete m_mixerChannelViews[index]->m_fader;
+		delete m_mixerChannelViews[index]->m_muteBtn;
+		delete m_mixerChannelViews[index]->m_soloBtn;
+		// delete mixerLine later to prevent a crash when deleting from context menu
+		m_mixerChannelViews[index]->m_mixerLine->hide();
+		m_mixerChannelViews[index]->m_mixerLine->deleteLater();
+		delete m_mixerChannelViews[index]->m_rackView;
+		delete m_mixerChannelViews[index];
+		m_channelAreaWidget->adjustSize();
+
+		// make sure every channel knows what index it is
+		for (int i = index + 1; i < m_mixerChannelViews.size(); ++i)
+		{
+			m_mixerChannelViews[i]->m_mixerLine->setChannelIndex(i - 1);
+		}
+		m_mixerChannelViews.remove(index);
+
+		// select the next channel
+		if (selLine >= m_mixerChannelViews.size()) 
+		{ 
+			selLine = m_mixerChannelViews.size() - 1; 
+		}
+		setCurrentMixerLine(selLine);
+
+		updateMaxChannelSelector();
 	}
-	m_mixerChannelViews.remove(index);
+}
 
-	// select the next channel
-	if( selLine >= m_mixerChannelViews.size() )
+bool MixerView::confirmRemoval(int index)
+{
+	bool needConfirm = ConfigManager::inst()->value("ui", "mixerchanneldeletionwarning", "1").toInt();
+	if (!needConfirm) { return true; }
+
+	TrackContainer::TrackList tracks;
+	tracks += Engine::getSong()->tracks();
+	tracks += Engine::patternStore()->tracks();
+
+	bool inUse = false;
```

Comment:
I think this check can be moved to a core function. `MixerView::deleteUnusedChannels` also contains a similar(but not identical) check, so it might also be a good candidate for refactoring.

Reply:
Ok. Created new function in Mixer.cpp and refactor deleteUnusedChannels accordingly.